### PR TITLE
Fix "Add unambiguous imports on the fly" in 223 platform

### PR DESCRIPTION
**Pull request by dima74**
changelog: Fix `Add unambiguous imports on the fly` for 2022.3

### CODE REVIEW

File: src/223/main/kotlin/org/rust/ide/inspections/import/RsReferenceImporter.kt [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import org.rust.lang.core.psi.ext.RsElement
+import org.rust.lang.core.psi.ext.containingCrate
+import org.rust.lang.core.psi.ext.isUnderCfgTest
+import java.util.function.BooleanSupplier
+
+class RsReferenceImporter : ReferenceImporter {
+    override fun autoImportReferenceAtCursor(editor: Editor, file: PsiFile): Boolean = false
+
+    @Suppress("UnstableApiUsage")
+    override fun computeAutoImportAtOffset(editor: Editor, file: PsiFile, offset: Int, allowCaretNearReference: Boolean): BooleanSupplier? {
```
Comment (Undin) [inline src/223/main/kotlin/org/rust/ide/inspections/import/RsReferenceImporter.kt:25]: Could you check a platform version when this API appeared and use it as minor supported version for 223 builds?

File: src/223/main/kotlin/org/rust/ide/inspections/import/RsReferenceImporter.kt [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        val reference = file.findReferenceAt(offset) ?: return null
+        val element = reference.element as? RsElement ?: return null
+
+        val context = when (element) {
+            is RsPath -> RsUnresolvedReferenceInspection.processPath(element)?.second
+            is RsMethodCall -> AutoImportFix.findApplicableContext(element)
```
Comment (Undin) [inline src/223/main/kotlin/org/rust/ide/inspections/import/RsReferenceImporter.kt:31]: It seems we need some common place for all auto import context API. Like we grouped all `import` methods in `RsImportHelper`. Probably, it's worth to extract `Context` to separate class with some proper name and move all methods to constuct it into its companion object
But I think it should be done in separate PR

File: src/223/main/kotlin/org/rust/ide/inspections/import/RsReferenceImporter.kt [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        val element = reference.element as? RsElement ?: return null
+
+        val context = when (element) {
+            is RsPath -> RsUnresolvedReferenceInspection.processPath(element)?.context
+            is RsMethodCall -> AutoImportFix.findApplicableContext(element)
+            else -> return null
+        }
+        if (context == null || context.candidates.size != 1) return null
+
+        if (element.containingCrate?.hasCyclicDevDependencies == true && element.isUnderCfgTest) return null
```
[the remark this answers is not in the document]
Comment (dima74) [reply src/223/main/kotlin/org/rust/ide/inspections/import/RsReferenceImporter.kt]: > Can we have only one place with auto import fix creation?

I don't see a simple way to do this, and I think this PR is not the best one for massive refactorings, so I propose extracting the duplicated code into common function

File: src/223/main/kotlin/org/rust/ide/inspections/import/RsReferenceImporter.kt [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+            is RsPath -> RsUnresolvedReferenceInspection.processPath(element)?.context
+            is RsMethodCall -> AutoImportFix.findApplicableContext(element)
+            else -> return null
+        }
+        if (context == null || context.candidates.size != 1) return null
+
+        if (element.shouldIgnoreUnresolvedReference()) return null
+
+        return BooleanSupplier {
+            AutoImportFix(element, context).invoke(element.project)
```
Comment (vlad20012) [inline src/223/main/kotlin/org/rust/ide/inspections/import/RsReferenceImporter.kt:38]: Direction action invocation leads to a separate entry in undo log, so Ctrl+Z undoes auto-import instead of actual typing that is extremely annoying:
https://user-images.githubusercontent.com/3221931/201402110-7511fd56-00f6-4857-a4ef-496b5e1fab6d.mp4

```suggestion
            runUndoTransparentWriteAction {
                context.candidates.first().import(element)
            }
```

Comment (dima74) [reply src/223/main/kotlin/org/rust/ide/inspections/import/RsReferenceImporter.kt:38]: I agree, but looks like `runUndoTransparentWriteAction` doesn't apply formatting (new import can be inserted without new line), so I propose fixing it in a separate PR (anyway this behaviour is present in 2022.2 as well)

Comment (vlad20012) [reply src/223/main/kotlin/org/rust/ide/inspections/import/RsReferenceImporter.kt:38]: Okay

**Review by vlad20012 — APPROVED** [review]
bors r+

File: src/223/main/kotlin/org/rust/ide/inspections/import/RsReferenceImporter.kt (hunk 4 above)
Comment (dima74) [reply src/223/main/kotlin/org/rust/ide/inspections/import/RsReferenceImporter.kt:38]: #9737
